pennylane/beta/tapes/qubit_param_shift.py

```diff
@@ -0,0 +1,187 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Qubit parameter shift quantum tape.
+
+Provides analytic differentiation for all one-parameter gates where the generator
+only has two unique eigenvalues; this includes one-parameter single-qubit gates.
+"""
+# pylint: disable=attribute-defined-outside-init
+import copy
+import numpy as np
+
+import pennylane as qml
+from pennylane.beta.queuing import MeasurementProcess
+
+from .tape import QuantumTape
+
+
+class QubitParamShiftTape(QuantumTape):
+    """Quantum tape for qubit parameter-shift analytic differentiation method.
+
+    For more details on the quantum tape, please see :class:`~.QuantumTape`.
+    """
+
+    def _update_circuit_info(self):
+        super()._update_circuit_info()
+
+        self.analytic_pd = self._parameter_shift
+
+        self.var_idx = None
+        self.var_mask = [m.return_type is qml.operation.Variance for m in self.measurements]
+        self._original_obs = self._obs.copy()
+
+        if any(self.var_mask):
+            self.analytic_pd = self._parameter_shift_var
+            self.var_idx = np.where(self.var_mask)[0]
+
+        # # expand out the tape, if any operations are not supported by the parameter shift rule
+        # if {"CRX", "CRZ", "CRY", "CRot"} & {op.name for op in self.operations}:
+        #     stop = set(device.operations) - {"CRX", "CRZ", "CRY", "CRot"}
+        #     self = self.expand(depth=2, stop_at=stop)
+
+
+    def _grad_method(self, idx, use_graph=True):
+        op = self._par_info[idx]["op"]
+
+        if op.grad_method is None:
+            return None
+
+        if op.grad_method == "F":
+            return "F"
+
+        if (self._graph is not None) or use_graph:
+            # an empty list to store the 'best' partial derivative method
+            # for each observable
+            best = []
+
+            # loop over all observables
+            for ob in self.observables:
+                # check if op is an ancestor of ob
+                has_path = self.graph.has_path(op, ob)
+
+                # Use analytic method if there is a path, else the gradient is zero
+                best.append("A" if has_path else "0")
+
+            if all(k == "0" for k in best):
+                return "0"
+
+        return "A"

```

Comment:
Could we make this have less duplication? e.g.
```python
def _grad_method(self, idx, use_graph=True):
    op = self._par_info[idx]["op"]

    if op.grad_method == "F":
            return "F"

    return super()._grad_method(idx, use_graph=use_graph, meth="A")
```

Reply:
good idea!